### ISSUE TITLE
VContainer/Message Pipe exception when building for iOS

### ISSUE THREAD

**Issue by Kevdome3000** [hadashiA/VContainer#253]
Hey there, 

Is CodeGen currently working? I saw the other issues regarding codeGen but couldn't find a definitive answer.

I'm using MessagePipe and all works well in editor but when I build for iOS and the app fails on intilization with the following errors.

I should also mention that I'm using Odin for AOT serialization as well. Possible that's the source of conflict?

```
VContainerException: MessagePipe.AsyncRequestHandler`2[[XXXXX.Domains.Messages.ICommand, XXXXX.Domain.Assembly, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null],[Application.Framework.Messaging.IResult, Application, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null]] is not assignable from MessagePipe.IAsyncRequestHandler`2[[XXXX.Domains.Messages.ICommand, Application.Domain.Assembly, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null],[Application.Framework.Messaging.IResult, Application, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null]]
  at VContainer.RegistrationBuilder.AddInterfaceType (System.Type interfaceType) [0x00000] in <00000000000000000000000000000000>:0 
  at VContainer.RegistrationBuilder.As (System.Type interfaceType) [0x00000] in <00000000000000000000000000000000>:0 
  at MessagePipe.ContainerBuilderExtensions.RegisterAsyncRequestHandler[TRequest,TResponse,THandler] (VContainer.IContainerBuilder builder, MessagePipe.MessagePipeOptions options) [0x00000] in <00000000000000000000000000000000>:0 
  at AppManager.RegisterMessagePipe (VContainer.IContainerBuilder builder) [0x00000] in <00000000000000000000000000000000>:0 
  at VContainer.Unity.LifetimeScope.InstallTo (VContainer.IContainerBuilder builder) [0x00000] in <00000000000000000000000000000000>:0 
  at VContainer.Unity.LifetimeScope.Build () [0x00000] in <00000000000000000000000000000000>:0 
  at VContainer.Unity.LifetimeScope.Awake () [0x00000] in <00000000000000000000000000000000>:0 

VContainerException: MessagePipe.AsyncRequestHandler`2[[Application.Domains.Messages.ICommand, Application.Domain.Assembly, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null],[Application.Framework.Messaging.IResult, Application, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null]] is not assignable from MessagePipe.IAsyncRequestHandler`2[[Application.Domains.Messages.ICommand, Application.Domain.Assembly, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null],[Application.Framework.Messaging.IResult, Application, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null]]
  at VContainer.RegistrationBuilder.AddInterfaceType (System.Type interfaceType) [0x00000] in <00000000000000000000000000000000>:0 
  at VContainer.RegistrationBuilder.As (System.Type interfaceType) [0x00000] in <00000000000000000000000000000000>:0 
  at MessagePipe.ContainerBuilderExtensions.RegisterAsyncRequestHandler[TRequest,TResponse,THandler] (VContainer.IContainerBuilder builder, MessagePipe.MessagePipeOptions options) [0x00000] in <00000000000000000000000000000000>:0 
  at FitMama.AppManager.RegisterMessagePipe (VContainer.IContainerBuilder builder) [0x00000] in <00000000000000000000000000000000>:0 
  at VContainer.Unity.LifetimeScope.InstallTo (VContainer.IContainerBuilder builder) [0x00000] in <00000000000000000000000000000000>:0 
  at VContainer.Unity.LifetimeScope.Build () [0x00000] in <00000000000000000000000000000000>:0 
  at VContainer.Unity.LifetimeScope.Awake () [0x00000] in <00000000000000000000000000000000>:0 
UnityEngine.UnhandledExceptionHandler:PrintException(String, Exception)
UnityEngine.UnhandledExceptionHandler:HandleUnhandledException(Object, UnhandledExceptionEventArgs)
System.UnhandledExceptionEventHandler:Invoke(Object, UnhandledExceptionEventArgs)
```

**Comment by hadashiA** [comment]
According to the stack trace, there is a run error when the Register* is executed.
( Perhaps it's not a CodeGen issue.

It looks like the following Validation is failed.
https://github.com/hadashiA/VContainer/blob/master/VContainer/Assets/VContainer/Runtime/RegistrationBuilder.cs#L109

In other words, the following seems to be `false`.
```csharp
typeof(IAsyncRequestHandler<ICommand, IResult>).IsAssignableFrom(typeof(AsyncRequestHandler<ICommand, IResult>))
```

However, I have not figured out why this only happens during iOS build 🤔 
I'll look into it later.

**Comment by Kevdome3000** [comment]
Thanks for looking into it!

I'll standby

**Comment by hadashiA** [comment]
@Kevdome3000  I tried an iOS build with the small sample below and did not encounter the same error.  🤔 
If you send me the code or project to reproduce, I'll  look into it a bit more. 

Note:
However, on the first try, I encountered another error (AsyncRequestHandler constructor not found).
So I put the `link.xml` file in Assets/ to work around it.
( It could possibly have something to do with the error you encountered.

```link.xml
<linker>
    <assembly fullname="MessagePipe" preserve="all"/>
</linker>
```

```csharp
public interface ICommand
{
    string Name { get; }
}

public interface IResult
{
}

public readonly struct CommandA : ICommand
{
    public string Name { get; }

    public CommandA(string name)
    {
        Name = name;
    }
}

public readonly struct ResultA : IResult
{
}

public class FooRequestHandler : IAsyncRequestHandler<ICommand, IResult>
{
    public UniTask<IResult> InvokeAsync(ICommand request, CancellationToken cancellationToken = new CancellationToken())
    {
        UnityEngine.Debug.Log($"FooRequestHandler.InvokeAsync : {request.Name}");
        return UniTask.FromResult((IResult)new ResultA());
    }
}

class SampleEntryPoint : IAsyncStartable
{
    readonly IAsyncRequestHandler<ICommand, IResult> requestHandler;

    public SampleEntryPoint(IAsyncRequestHandler<ICommand, IResult> requestHandler)
    {
        this.requestHandler = requestHandler;
    }

    public async UniTask StartAsync(CancellationToken cancellation)
    {
        UnityEngine.Debug.Log("SampleEntryPoint.StartAsync");
        await requestHandler.InvokeAsync(new CommandA("AAAAA"), cancellation);
    }
}

public class BaseLifetimeScope : LifetimeScope
{
    protected override void Configure(IContainerBuilder builder)
    {
            builder.RegisterEntryPoint<SampleEntryPoint>();

            var options = builder.RegisterMessagePipe();
            builder.RegisterAsyncRequestHandler<ICommand, IResult, FooRequestHandler>(options);
    }
}
```

**Comment by Kevdome3000** [comment]
@hadashiA It looks like the link file did the trick , the build worked as expected. 

Such an annoyance🙄 but thank you again for taking the time to look into the issue and also many thanks for the awesome library!

**Comment by hadashiA** [comment]
Ok, that's good.

About the IL2CPP strip...
MessagePipe has been merged with a fix to `[Preserve]` the AsyncRequestHandler.
I think link.xml is no longer needed in this version.  😃 
https://github.com/Cysharp/MessagePipe/pull/53

**Comment by neuecc** [comment]
Thanks for the investigation.
I've released 1.5.1, it includes hadashiA's fantastic fix.
